partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/RexConverter.kt

```diff
@@ -148,73 +132,57 @@ internal object RexConverter {
                 isMissingValue = node.lit.code() == Literal.MISSING
             )
-            val op = rexOpLit(pValue)
+            val op = rexOpLit(datum)
             return rex(cType, op)
         }
 
-        private fun Literal.toPartiQLValue(): PartiQLValue {
+        private fun Literal.toDatum(): Datum {
             val lit = this
             return when (lit.code()) {
-                Literal.NULL -> nullValue()
-                Literal.MISSING -> missingValue()
-                Literal.STRING -> stringValue(lit.stringValue())
-                Literal.BOOL -> boolValue(lit.booleanValue())
+                Literal.NULL -> Datum.nullValue()
+                Literal.MISSING -> Datum.missing()
+                Literal.STRING -> Datum.string(lit.stringValue())
+                Literal.BOOL -> Datum.bool(lit.booleanValue())
                 Literal.EXACT_NUM -> {
-                    // TODO previous behavior inferred decimals with scale = 0 to be a PartiQLValue.IntValue with
-                    //  PType of numeric. Since we're keeping numeric and decimal, need to take another look at
-                    //  whether the literal should have type decimal or numeric.
                     val dec = lit.bigDecimalValue().round(MathContext(38, RoundingMode.HALF_EVEN))
-                    if (dec.scale() == 0) {
-                        intValue(dec.toBigInteger())
-                    } else {
-                        decimalValue(dec)
-                    }
+                    Datum.decimal(dec, dec.precision(), dec.scale())
                 }
                 Literal.INT_NUM -> {
                     val n = lit.numberValue()
                     // 1st, try parse as int
                     try {
                         val v = n.toInt(10)
-                        return int32Value(v)
+                        return Datum.integer(v)
                     } catch (ex: NumberFormatException) {
                         // ignore
                     }
 
                     // 2nd, try parse as long
                     try {
                         val v = n.toLong(10)
-                        return int64Value(v)
+                        return Datum.bigint(v)
                     } catch (ex: NumberFormatException) {
                         // ignore
                     }
 
                     // 3rd, try parse as BigInteger
                     try {
                         val v = BigInteger(n)
-                        return intValue(v)
+                        val vDecimal = BigDecimal(v)
+                        return Datum.decimal(vDecimal, vDecimal.precision(), vDecimal.scale())
                     } catch (ex: NumberFormatException) {
                         throw ex
                     }
                 }
                 Literal.APPROX_NUM -> {
-                    float64Value(lit.numberValue().toDouble())
+                    return Datum.doublePrecision(lit.numberValue().toDouble())
```

Comment:
More so a question for the plan representation: are we committing to using `Datum` as the literal representation in the plan? I haven't looked much into serialization of plans but not sure if eagerly converting some of the types into Java can lead to some later problems.

At least for `VARIANT`, we don't appear to be any `lower`ing till evaluation time so wondering if the same can apply for literals in general.

Reply:
The plan representation of literals is out of scope of this PR -- so I'm not sure.

And, what do you mean by the second observation?

Reply:
This PR is (internal) plan representation of literals since the definition of literals in plans has been changed by this,

```
    - partiql_value::'org.partiql.value.PartiQLValue',
    + partiql_value::'org.partiql.spi.value.Datum',
```

But I think you two might be talking about _serialization_ of values which is covered by "PartiQL encoding in Ion" in #1655